# Import results

Import functions allow the users to import a `JSON` file (extension `.json`), or multiple files, with their own plasmid data.
These files are generated by [pATLASflow](https://github.com/tiagofilipe12/pATLASflow), a pipeline to run mapping, 
mash screen and assembly methods for pATLAS. 

## Download and install requirements to run the pipeline

**[pATLASflow](https://github.com/tiagofilipe12/pATLASflow) is a [NextFlow](https://www.nextflow.io/)
pipeline.**

### Requirements

* [Java 8](http://www.oracle.com/technetwork/java/javase/downloads/index.html) or higher.
* [Docker](https://docs.docker.com/install/) or [Singularity](http://singularity.lbl.gov/install-linux).
* [Nextflow](https://www.nextflow.io/docs/latest/getstarted.html#installation) 

#### Conda recipe for nextflow

Nextflow can be installed through bioconda: [![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square)](http://bioconda.github.io/recipes/nextflow/README.html)

```
conda install nextflow
```

## Mapping

The mapping pipeline can be run with the following command: 

`nextflow run tiagofilipe12/pATLASflow --mapping --reads "your_folder/*.fastq"`

The resulting `JSON` file can then be provided to pATLAS in the **Mapping** menu.

## Mash screen

The mash screen pipeline can be run with the following command:

`nextflow run tiagofilipe12/pATLASflow --mash_screen --reads "your_folder/*.fastq"`

The resulting `JSON` file can then be provided to pATLAS in the **Mash screen** menu.

## Sequence

The sequence pipeline can be run with the following command:

`nextflow run tiagofilipe12/pATLASflow --assembly --fasta "your_folder/*.fasta"`

The resulting `JSON` file can then be provided to pATLAS in the **Sequence** menu.